include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_then_scan.h

```diff
@@ -0,0 +1,843 @@
+                __v = __gen_input(__in_rng, __local_id);
+                __sub_group_scan_partial<__sub_group_size, __is_inclusive, /*__init_present=*/true>(
+                    __sub_group, __scan_input_transform(__v), __binary_op, __sub_group_carry,
+                    __n - (__subgroup_start_id + (__iters - 1) * __sub_group_size));
+                if constexpr (__capture_output)
+                {
+                    if (__offset < __n)
+                        __write_op(__out_rng, __offset, __v);
+                }
+            }
+        }
+    }
+}
+
+template <typename... _Name>
+class __reduce_then_scan_reduce_kernel;
+
+template <typename... _Name>
+class __reduce_then_scan_scan_kernel;
+
+template <std::size_t __sub_group_size, std::size_t __max_inputs_per_item, bool __is_inclusive,
+          typename _GenReduceInput, typename _ReduceOp, typename _InitType, typename _KernelName>
+struct __parallel_reduce_then_scan_reduce_submitter;
+
+template <std::size_t __sub_group_size, std::size_t __max_inputs_per_item, bool __is_inclusive,
+          typename _GenReduceInput, typename _ReduceOp, typename _InitType, typename... _KernelName>
+struct __parallel_reduce_then_scan_reduce_submitter<__sub_group_size, __max_inputs_per_item, __is_inclusive,
+                                                    _GenReduceInput, _ReduceOp, _InitType,
+                                                    __internal::__optional_kernel_name<_KernelName...>>
+{
+    // Step 1 - SubGroupReduce is expected to perform sub-group reductions to global memory
+    // input buffer
+    template <typename _ExecutionPolicy, typename _InRng, typename _TmpStorageAcc>
+    sycl::event
+    operator()(_ExecutionPolicy&& __exec, const sycl::nd_range<1> __nd_range, _InRng&& __in_rng,
+               _TmpStorageAcc& __scratch_container, const sycl::event& __prior_event,
+               const std::size_t __inputs_per_sub_group, const std::size_t __inputs_per_item,
+               const std::size_t __block_num) const
+    {
+        using _InitValueType = typename _InitType::__value_type;
+        return __exec.queue().submit([&, this](sycl::handler& __cgh) {
+            __dpl_sycl::__local_accessor<_InitValueType> __sub_group_partials(__num_sub_groups_local, __cgh);
+            __cgh.depends_on(__prior_event);
+            oneapi::dpl::__ranges::__require_access(__cgh, __in_rng);
+            auto __temp_acc = __scratch_container.__get_scratch_acc(__cgh);
+            __cgh.parallel_for<_KernelName...>(
+                    __nd_range, [=, *this](sycl::nd_item<1> __ndi) [[sycl::reqd_sub_group_size(__sub_group_size)]] {
+                _InitValueType* __temp_ptr = _TmpStorageAcc::__get_usm_or_buffer_accessor_ptr(__temp_acc);
+                std::size_t __group_id = __ndi.get_group(0);
+                __dpl_sycl::__sub_group __sub_group = __ndi.get_sub_group();
+                std::uint32_t __sub_group_id = __sub_group.get_group_linear_id();
```

Comment:
For a sub_group, the return type is `uint32_t` rather than `size_t` from the sycl spec, better to use a smaller type when we can.
We may be able to shrink a few other types, but this one is clear.